featuretools/synthesis/deep_feature_synthesis.py

```diff
@@ -1116,43 +1159,34 @@ def can_stack_primitive_on_inputs(primitive, inputs):
         # However, we want to check stacking rules with the primitive the DirectFeature is based on.
         f_primitive = _find_root_primitive(feature)
 
-        if not primitive_stack_on_self and isinstance(f_primitive, primitive_class):
-            return False
-
-        if isinstance(f_primitive, tup_primitive_stack_on_exclude):
-            return False
-
-        if feature.number_output_features > 1:
-            return False
-
-        if f_primitive.base_of_exclude is not None and isinstance(
+        # check if stacking is disabled
+        if _check_if_stacking_is_prohibited(
+            feature,
+            f_primitive,
             primitive,
-            tuple(f_primitive.base_of_exclude),
+            primitive_class,
+            primitive_stack_on_self,
+            tuple_primitive_stack_on_exclude,
         ):
             return False
 
-        if primitive_stack_on_self and isinstance(f_primitive, primitive_class):
-            continue
-
-        if tup_primitive_stack_on is None or isinstance(
+        # we permit stacking only if it is not prohibited and meets the criterion to be permitted
+        if _check_if_stacking_is_permitted(
```

Comment:
maybe I'm missing something, but prohibited and permitted should be mutually exclusive states, and so if its not prohibited, its permitted. Can you refactor even more to make this the case?

Reply:
@dvreed77 I don't think prohibited and permitted in this case are mutually exclusive. I think we can have cases where both checks are `False` or both are `True` because the logic checks are different between the "permitted" and "prohibited" functions. We want prohibited to take precedence, so in the case that both checks are True or both checks are False, we want the result to be "prohibited".

I might be missing something still, but I think those conditions make it a little tricker to refactor this further.

Reply:
I did add some additional comments to maybe clarify what is going on a little better. I'm open to other suggestions on how to improve further though.

Reply:
I agree with Nate here, I remember trying to find a way to clean it up before coming up with the original changes, but none were immediately apparent to me at the time. I think this PR was an initial attempt at refactoring the stacking logic at a surface level, mostly to play around with and discover the areas where we could refactor. A second, more fundamental refactor might be needed to simplify the deeper logic of how we handle stacking.